app/controllers/build_controller.rb

```diff
@@ -0,0 +1,155 @@
+    products.each do |product|
+      product['product_schema'] = JSON.parse(product['product_schema'])
+    end
+  end
+
+
+  def get_build(product_categories, requirements)
+    if requirements
+      product_categories = filter(product_categories, requirements)
+    end
+
+    first_pick_category = Rails.configuration.build_first_component
+    product_categories.each do |product_category|
+      if first_pick_category ==  product_category['category_name']
+        motherboard = product_category['products'].first
+        self.components[first_pick_category] = motherboard
+      end
+    end
+
+    product_categories.each do |product_category|
+      category = product_category['category_name']
+      products = product_category['products']
+
+      if self.components[category].nil?
+        stuck_at_category = 0
+        products.each do |product|
+          compatible, not_compatible_at_category = is_compatible(category, product)
+
+          if compatible
+            self.components[category] = product
+            stuck_at_category = nil
+            break
+          else
+            stuck_at_category = not_compatible_at_category
+          end
+        end
+
+        if stuck_at_category
+          product_categories = remove_product(stuck_at_category, self.components[stuck_at_category], product_categories)
+          self.components.delete(stuck_at_category)
+
+          get_build(product_categories, nil)
+        end
+      end
+    end
+
+    if self.components.size == self.products.size
+      self.components
+    else
+      nil
+    end
+  end
+
+  def is_compatible(category, product)
+    components = self.components
+    if components.empty?
+      return true, nil
+    else
+      if category == 'motherboard'
+
+        if components['memory']
+          if components['memory']['memory_type'] != product['memory_type']
+            return false, 'memory'
+          end
+        end
+
+        if components['case']
+          if components['case']['form_factor'] != product['Moederbord formaat']
+            return false, 'case'
+          end
+        end
+
+        if components['cpu']
+          if components['cpu']['socket'] != product['socket']
+            return false, 'cpu'
+          end
+        end
+      end
+
+      if category == 'memory' && components['motherboard']
+        if components['motherboard']['memory_type'] != product['memory_type']
+          return false, 'motherboard'
+        end
+      end
+
+      if category == 'case' && components['motherboard']
+        if components['motherboard']['form_factor'] != product['Moederbord formaat']
+          return false, 'motherboard'
+        end
+      end
+
+      if category == 'cpu' && components['motherboard']
+        if components['motherboard']['socket'] != product['socket']
+          return false, 'motherboard'
+        end
+      end
+
+      # if category == 'video_card' && components.has_key?('motherboard')
```

Comment:
Ja, ik laat videokaart achterwegen aangezien niet alle velden aanwezig zijn en we hogere prioriteiten hebben denk ik?